### PR TITLE
Add M20 L option to include long filenames in the listing.

### PR DESCRIPTION
**Pull request by mh-dm**
Updates M20 command to match docs at https://marlinfw.org/docs/gcode/M020.html
Aim was to have the absolute minimal implementation to both match the
docs and to support applications like OctoPrint that already have
support for this option and format as per https://github.com/OctoPrint/OctoPrint/issues/1600

Differences for M20 itself in cardreader.cpp printListing():
- optimization: dosFilename is no longer copied and stored twice
- no more leading `/` to match docs: `/SHORTDIR/BED_LE~1.GCO` -> `SHORTDIR/BED_LE~1.GCO`

*Notes*
- long name are (still) not used in selecting the file to print.
- feature is guarded by `LONG_FILENAME_HOST_SUPPORT` as per docs.
- kept `???` output for files/folders without long filenames for consistency with `M33`.

### Related Issues
#11403 #19091 #19111 #19960
I've done a thorough search but could not find already existing active issues or pull requests. Hope I'm not overstepping.

Looked at pull/19977 for inspiration so quick thanks to @RFBomb and @thinkyhead.

Example output:
````
M20
Begin file list
KFACTO~1.GCO 5752
100KB.GCO 102400
TEST_T~1.GCO 185094
SHORTDIR/BED_LE~1.GCO 666
SHORTDIR/MULTI/BED_LE~1.GCO 666
LONGDI~1/BED_LE~1.GCO 666
LONGDI~1/MULTIL~1/BED_LE~1.GCO 666
End file list
ok

M20 L
Begin file list
KFACTO~1.GCO 5752 kfactor.gcode
100KB.GCO 102400 ???
TEST_T~1.GCO 185094 test_tower.gcode
SHORTDIR/BED_LE~1.GCO 666 shortDir/bed_level.gcode
SHORTDIR/MULTI/BED_LE~1.GCO 666 shortDir/multi/bed_level.gcode
LONGDI~1/BED_LE~1.GCO 666 longDirName/bed_level.gcode
LONGDI~1/MULTIL~1/BED_LE~1.GCO 666 longDirName/multiLong/bed_level.gcode
End file list
ok

### CODE REVIEW

**Comment by mh-dm** [comment]
Tested by sending M20 and M20 L commands to printer (Ender-3 with SKR 1.3) both with LONG_FILENAME_HOST_SUPPORT enabled and disabled (in which case the L does nothing).

File: Marlin/src/sd/cardreader.cpp [hunk 1]
```diff
@@ -337,10 +340,10 @@ void CardReader::printListing(
 //
 // List all files on the SD card
 //
-void CardReader::ls(bool includeLongNames) {
+void CardReader::ls(TERN_(LONG_FILENAME_HOST_SUPPORT, bool includeLongNames/*=false*/)) {
   if (flag.mounted) {
     root.rewind();
-    printListing(root, includeLongNames);
+    printListing(root, nullptr OPTARG(LONG_FILENAME_HOST_SUPPORT, includeLongNames));
```
Comment (mihai-dumitresq) [inline Marlin/src/sd/cardreader.cpp:346]: Not sure if I'm reading the error right but: Remove nullptr?

Comment (mihai-dumitresq) [reply Marlin/src/sd/cardreader.cpp:346]: Nevermind, not correct.

**Comment by mh-dm** [comment]
Looks all green. I've also tested M20 with and without L after latest changes. Anything else I should do?

**Comment by mh-dm** [comment]
@thinkyhead Can you please be the reviewer for this or assign a reviewer?

**Comment by thinkyhead** [comment]
> Can you please be the reviewer for this or assign a reviewer?

Are you on a tight deadline? ;-)

**Comment by thinkyhead** [comment]
Tested, and it works! The long path is appended after the file size, as the host expects. Thanks for the contribution. If there are any other supported alternative formats (file size last?) we can also try those.

I had started on `G20 L` previously, but also started experimentation with a compact hierarchical JSON format listing (like RepRapFirmware) which would be more efficient, and some other commands to permit file operations more like a shell. I will return to this stuff again as part of #10610.
```
 * Extended interface:
 *   M20   : Deep listing with DOS8.3 path and file size.
 *   M20 L : Deep listing with longname path and file size.
 *   M20 H : Hierarchical (compact) listing with DOS paths and file sizes.
 *   M20 D : List the contents of the current DIR only, no prefix. With L show long names too.
 *   M20 C : Change the workDir to the given directory (long or short name).
 *   M20 U : Up one level.
 *   M20 R : Change workDir to the root directory.
 *   M20 W : What is the working directory?
```

**Comment by mh-dm** [comment]
> Are you on a tight deadline? ;-)

Haha, thankfully no.

> Tested, and it works! The long path is appended after the file size, as the host expects. Thanks for the contribution. If there are any other supported alternative formats (file size last?) we can also try those.

Thank you! I'm not looking to add other formats but I will look into adding files with long filenames, maybe through an extra L parameter to M28 Start SD write. Would be very useful for completely remotely managing SD files from octoprint or similar (though 8.3 filenames will still be authoritative). I'll prototype, hopefully get a small patch, then initiate a discussion. If you were already looking to add something like that and would like to offload, I could help by starting on the prototype sooner.

**Comment by thinkyhead** [comment]
> maybe through an extra L parameter to M28 Start SD write

Unfortunately at the time when Marlin was written the SDFat library did not have long filename support, so SDFat was copied into Marlin and some minimal long filename support was added on, along with other optimizations like caching with alpha sorting. Therefore, the SDFat code in `Marlin/src/sd` doesn't yet have the capability to create long filename entries. The [SDFat library](https://github.com/greiman/SdFat) did eventually gain long filename support, but the new library became too divergent to be compatible with Marlin, and it was never adopted. So, we will need to add new code to our own SDFat library if we want to create extended / long filename entries.

**Comment by Will-wastelander** [comment]
> > maybe through an extra L parameter to M28 Start SD write
> 
> Unfortunately at the time when Marlin was written the SDFat library did not have long filename support, so SDFat was copied into Marlin and some minimal long filename support was added on, along with other optimizations like caching with alpha sorting. Therefore, the SDFat code in `Marlin/src/sd` doesn't yet have the capability to create long filename entries. The [SDFat library](https://github.com/greiman/SdFat) did eventually gain long filename support, but the new library became too divergent to be compatible with Marlin, and it was never adopted. So, we will need to add new code to our own SDFat library if we want to create extended / long filename entries.

Is this doable ? It's really helpful when it comes to Creality v4.x.x boards, and with MKS Robin boards. It's possible that others would benefit from this also.

**Comment by thinkyhead** [comment]
> Is this doable?

It's possible we can retain our current `SDFat`-based code for the benefit of smaller AVR boards while migrating to something new for other HALs. I'll look into this more as I work with BTT and other contributors to update the "btt_ui" code. It's tricker now due to the added complexities of multi-volume support, but we'll keep chipping away.

**Comment by foosel** [comment]
Is there a reason why this was encoded in the capability flag `LONG_FILENAME` instead of `EXTENDED_M20` as documented on the [GCODE wiki node](https://reprap.org/wiki/G-code#M20:_List_SD_card) and already implemented in PrusaMarlin? I stuck to the only "spec" we have for OctoPrint's support of this and just got informed that this now won't work with mainline Marlin.

**Comment by thisiskeithb** [comment]
> I stuck to the only "spec" we have for OctoPrint's support of this and just got informed that this now won't work with mainline Marlin.

This will be fixed in https://github.com/MarlinFirmware/Marlin/pull/22941

**Comment by foosel** [comment]
Thank you, that's great to hear!

**Pull request merged**